### ISSUE TITLE
Inconsistent/Incorrect results

### ISSUE THREAD

**Issue by mlomb** [pykeio/ort#108]
Hi, first of all thanks for creating `ort`!

I'm trying to run inference of a very simple model but get drastically different (incorrect) results from onnxruntime in Python. If I run inference with all zeroes in Python I get `[[0.5, 0.5]]` every time.

<details>
<summary>Model generation and reference (keras, tf2onnx and onnxruntime)</summary>

You can download the model here: [onnx_model.zip](https://github.com/pykeio/ort/files/13271878/onnx_model.zip)

```python
from tensorflow.keras import losses
from tensorflow.keras import Input, Model
from tensorflow.keras.layers import Dense, Flatten

import onnx
import onnxruntime
import tf2onnx
import numpy as np

input = Input(shape=(7,6,2,2), name='input')

x = Flatten()(input)
x = Dense(128, activation='relu')(x)
x = Dense(64, activation='relu')(x)
x = Dense(32, activation='relu')(x)
x = Dense(16, activation='relu')(x)
x = Dense(2, activation='softmax', name='output')(x)

model = Model(inputs=[input], outputs=[x])
model.compile(loss=[losses.CategoricalCrossentropy()], optimizer='adam', metrics=['accuracy'])
model.summary()

onnx_model, _ = tf2onnx.convert.from_keras(model)
onnx.save_model(onnx_model, f"onnx_model.onnx")

sess = onnxruntime.InferenceSession(f"onnx_model.onnx")
input_name = sess.get_inputs()[0].name
label_name = sess.get_outputs()[0].name

for i in range(500):
    input = np.zeros((1,7,6,2,2)).astype(np.float32)
    pred_onnx = sess.run([label_name], {input_name: input})[0]
    print(pred_onnx) # [[0.5 0.5]]
```
</details>

I'm using branch `v2`, and tried running the model in a loop, but get weird results which vary with the value I use in `with_intra_threads`:

<details>
<summary>Example with v2 branch</summary>

```rust
use ndarray::*;
use ort::{Environment, ExecutionProvider, SessionBuilder, Value};

fn main() -> ort::Result<()> {
    tracing_subscriber::fmt::init();

    let environment = Environment::builder()
        .with_execution_providers([ExecutionProvider::CPU(Default::default())])
        .build()?
        .into_arc();

    let session = SessionBuilder::new(&environment)?
        .with_intra_threads(8)?
        .with_model_from_file("onnx_model.onnx")?;

    loop {
        let input = Array5::<f32>::zeros((1, 7, 6, 2, 2));
        let output = Array2::<f32>::zeros((1, 2));

        let in_value = Value::from_array(input)?;
        let out_value = Value::from_array(output)?;

        let mut binding = session.create_binding()?;
        binding.bind_input("input", in_value)?;
        binding.bind_output("output", out_value)?;
        let outputs = binding.run()?;

        let data = outputs[0].extract_tensor::<f32>()?.view().t().into_owned();

        println!("{:?}", data.as_slice().unwrap());
    }
}
```

</details>

With `.with_intra_threads(1)`, I never get something close to `[0.5, 0.5]`:
```
[3.0262524e29, 0.501777]
[6.2462477e28, 5.76e-43]
[3.0262524e29, 4.1777e-41]
[6.246444e28, 5.76e-43]
[6.2462175e28, 5.76e-43]
[6.246172e28, 5.76e-43]
[6.246852e28, 5.76e-43]
[3.0262524e29, 0.501777]
[6.246444e28, 5.76e-43]
[6.246535e28, 5.76e-43]
[3.0262524e29, 2.7915669e23]
[6.246157e28, 5.76e-43]
[3.0262524e29, 0.501777]
[6.2460966e28, 5.76e-43]
```

With `.with_intra_threads(8)`, I _sometimes_ get the expected result:
```
[0.5, 0.5]
[0.5, 0.5]
[3.0262524e29, 0.501777]
[3.4692473e27, 2.7908747e23]
[8.391229e34, 7.67e-43]
[8.392116e34, 7.67e-43]
[0.5, 0.5]
[0.5, 0.5]
[0.5, 0.5]
[0.5, 0.5]
[0.5, 0.5]
[0.5, 0.5]
[8.392338e34, 7.67e-43]
[0.5, 0.5]
```

With CUDA, is the same story (affected by threads too) but with more incorrect results than correct.

I'm not sure if I'm messing up the lifetime of the tensors or something. I tried it in two different computers and faced the same issue, so I decided to report it. Hope it can be replicated.

-------

I also tried with the latest version (`1.16.2`). Each execution (session) gives a consistent result, but not across sessions. I came across: `[[0.5], [0.5]]`, `[[1], [0]]` (most likely), `[[0.57364064], [0.4263594]]` (or close) and sometimes even `[[NaN], [NaN]]`.

<details>
<summary>Example with 1.16.2</summary>

```rust
use ndarray::*;
use ort::{Environment, ExecutionProvider, SessionBuilder, Value, OrtResult, AllocatorType, MemType, MemoryInfo, AllocationDevice};

fn main() -> OrtResult<()> {
    tracing_subscriber::fmt::init();

    let environment = Environment::builder()
        .with_execution_providers([ExecutionProvider::CPU(Default::default())])
        .build()?
        .into_arc();

    let session = SessionBuilder::new(&environment)?
        .with_intra_threads(1)?
        .with_model_from_file("onnx_model.onnx")?;

    loop {
        let input = CowArray::from(Array5::<f32>::zeros((1, 7, 6, 2, 2)).into_dyn());

        let in_value = Value::from_array(session.allocator(), &input)?;
        let output_mem_info = MemoryInfo::new(AllocationDevice::CPU, 0, AllocatorType::Device, MemType::Default)?;

        let mut binding = session.bind()?;
        binding.bind_input("input", in_value)?;
        binding.bind_output("output", output_mem_info)?;

        session.run_with_binding(&binding)?;

        let outputs = binding.outputs()?;
        for (output_name, output_value) in outputs {
            let output_array = output_value.try_extract::<f32>()?;
            println!("{:}", output_array.view().t().into_owned());
        }
    }
}
```
</details>

------

Thanks!

**Comment by decahedron1** [comment]
With the v2 code on Windows I was getting a `STATUS_HEAP_CORRUPTION` exit code, which I fixed by removing the line to register the CPU execution provider. I'm able to reproduce the accuracy issues with `IoBinding`.

<details>
<summary>Foregoing `IoBinding` gives consistent `0.5`s.</summary>

```rs
use ndarray::*;
use ort::{Environment, ExecutionProvider, SessionBuilder};

fn main() -> ort::Result<()> {
	tracing_subscriber::fmt::init();

	let environment = Environment::builder().build()?.into_arc();

	let session = SessionBuilder::new(&environment)?
		.with_intra_threads(8)?
		.with_model_from_file("onnx_model.onnx")?;

	loop {
		let input = Array5::<f32>::zeros((1, 7, 6, 2, 2));

		let outputs = session.run(ort::inputs![input]?)?;

		let data = outputs[0].extract_tensor::<f32>()?.view().t().into_owned();

		println!("{:?}", data.as_slice().unwrap());
	}
}
```
</details>

**Comment by decahedron1** [comment]
For the segfault it seems like `OrtSessionOptionsAppendExecutionProvider_CPU` was silently removed in ONNX Runtime 1.16. I will backport the fix to `ort` 1.16 shortly.

As for the `IoBinding` issue, the output values are moved into the `IoBinding` in `bind_output` by adding them to a vector of `values`:
https://github.com/pykeio/ort/blob/84add5f0d22c36ed251e063e40a1457ea01ece69/src/io_binding.rs#L44-L50

When an `IoBinding` is run, it drops all values owned by the binding, including output values, **before then returning those outputs**:
https://github.com/pykeio/ort/blob/84add5f0d22c36ed251e063e40a1457ea01ece69/src/io_binding.rs#L60-L65

which is obviously going to cause an issue :^)

Changing `bind_output` to instead take a `&mut Value` is more ergonomic and fixes the lifetime issues, and I'm now getting consistent `0.5`s with & without `IoBinding`. Unfortunately, this is a breaking change, so I cannot backport it to 1.16.

Thank you for reporting this issue! Potential lifetime & ergonomics issues with `IoBinding` was one of my few remaining concerns about v2. Finding and fixing this issue is going to help its release move much more swiftly.

**Comment by mlomb** [comment]
Thanks for looking at this so fast! Glad it helped!
